rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+The download jobs are used by the upload service to know where to download the file from before uploading them to S3.
+
+##### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - the link to download. Use to retrieve the file from CSP.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- file_name - the name of the file being downloaded
+
+#### Upload Service
+
+A computing resource within the AWS cloud. It will receive download jobs from the Download Queue. Those jobs contain
+the information need by the compute resource to download a file from DropBox and upload it to S3. The computer resource
+will also verify the integrity of the download and upload. The compute resource will also change the state of the
+upload in the Upload table as needed.
+
+While the download job is processing, the computing resource with occasionally poll the upload table for a cancel
+pending status. If the cancel pending status is set, the computing resource will delete any data uploaded to S3,
+remove the download job from the queue, and change the status of the upload to canceled.
+
+If an error occurs while processing the upload, the compute resource will return the download job to the queue and
+update the upload table.
+
+##### Upload Table
+
+The upload table is a DynamoDB table that tracks the current status of uploads. The table will have the following columns:
+
+- dataset_id - primary key
+- status - the current state of the upload.
+- progress - the current progress of the upload.
+- owner - the id of the user who owns this collection.
+
+#### Upload Bucket
+
+The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
+before moving to the main bucket. Uploading files to a staging buckets makes it easier to clean up canceled, partial,
+or otherwise abandoned uploads, without accidentally deleting files actively being served by the Data Portal. Central
+Infra also provides a [bucket template](https://github.com/chanzuckerberg/cztack/tree/main/aws-s3-private-bucket) to help
+manage stale multipart uploads.
+
+S3 event can be used to trigger additional processing such as validation, once the upload has completed. The additional
+processing is outside the scope of this RFC.
+
+### Data Portal APIs
+
+The DP Browser App will use these endpoints to interact with the Backend using these APIs. This section describes the new
+endpoints and how they are used. These are the new API endpoints needed to complete an upload.
+
+These APIs replace the [PUT /v1/submission/{project_uuid}/file](https://docs.google.com/document/d/1d8tv2Ub5b3E7Il85adOAUcG8P05N6UBZJ3XbhJSRrFs/edit#heading=h.3hln6w2kzoyt)
+section.
+
+#### PUT submission/{submission_id}/upload/link
```

Comment:
I see that we have `DELETE submission/{submission_id}/upload` but this one is `PUT submission/{submission_id}/upload/link`

Does it help to have consistent URL:

`PUT submission/{submission_id}/upload` with `link` as part of the payload?

Or are we reserving `PUT submission/{submission_id}/upload` for uploading file through the browser?

Thanks!

Reply:
You got it. I plan on using `PUT submission/{submission_id}/upload/file` for uploading a file. `DELETE submission/{submission_id}/upload` would be used for either local or cloud.